### ISSUE TITLE
Docker image for making TLA+ tooling easily accessible

### ISSUE THREAD

**Issue by twitu** [tlaplus/tlaplus#683]
I've followed the [TLA+ video course](http://lamport.azurewebsites.net/video/videos.html) and gotten a chance to attend a TLA+ workshop conducted by @lemmy.

The video course recommended installing the standard TLA+ toolbox which was good for starting off. But the codespaces available in the [tlaplus-workshops/ewd998](https://github.com/tlaplus-workshops/ewd998) for was a much better experience and the tooling was much more accessible.

It would be great to have a Docker image with all the installations present in the above codespace. Combined with vs code's feature to use a [docker container as a development environment](https://code.visualstudio.com/docs/remote/create-dev-container) will make the well integrated tooling of TLA+ easily and locally accessible.

A possible output from this issue could just be a docker file or a docker image hosted in some registry, with a section in the README mentioning how it can be used.

**Comment by lemmy** [comment]
I would be good to know how common docker is among TLA+ users.  I don't have docker on my computers.

Perhaps, as a first step towards making the tools more accessible, we could try and make the TLA+ VSCode extension self-contained. If the vscode extension would bundle a Java VM -- like the Toolbox does -- a user would only have to install the extension in VSCode.

**Comment by twitu** [comment]
> FWIW: https://github.com/tlaplus-workshops/ewd998/blob/main/.devcontainer/install.sh is what Codespaces runs at startup on top of a recent Ubuntu Linux instance.

This is pretty helpful, it could be easily converted into a Dockerfile. A docker image would make the features installed by the script available in a platform independent manner.

> a first step towards making the tools more accessible, we could try and make the TLA+ VSCode extension self-contained

This would be an excellent upgrade. Admittedly it does leave users who prefer a different editor/IDE lacking. On the other hand, I personally like using VS Code and would find this solution quite helpful.

**Comment by lemmy** [comment]
@twitu Could you provide such a Dockerfile? :-)

**Comment by twitu** [comment]
Yes I'll take a look at this. Thanks for sharing this reference docker file - https://github.com/tlaplus/Examples/blob/master/.gitpod.Dockerfile.

My only concern is that it derives from gitpod/workspace-full-vnc, so I'm not sure how heavy it is or how well it works locally. I'll try it out and come back on this.

**Comment by lemmy** [comment]
The reason it derives from `gitpod/workspace-full-vnc` is to run the Eclipse-based TLA+ Toolbox in the browser via VNC.  With the VSCode TLA+ extension alone, a more basic image would work fine.

**Comment by kevinsullivan** [comment]
Help is here. Here's a good Ubuntu 22.04-based VSCode dev container for TLA+ including LaTeX pretty printing and display, auto-reparsing on file save, and state machine visualization using graphviz. https://github.com/kevinsullivan/TLAPlusDocker

**Comment by thesammiller** [comment]
There's a a Docker-based browser interface called WebGME with FORMULA as the model checker on the backend. The value to this approach is that instead of lifting the software into the cloud/JavaScript, the server can run the program locally as a process. There is the ability to exchange backends with other model checkers, although the front-end is not a text editor and more for UML-style diagrams with transitions that get translated into the model language. 

There's a paper describing the integration here: https://ceur-ws.org/Vol-2019/gemoc_1.pdf

This could possibly be a starting point for a web editor, although TLA+ would need to either conform to the GUI or implement another one.

**Comment by lemmy** [comment]
Feel free to reopen.
